ddtrace/internal/module.py

```diff
@@ -0,0 +1,430 @@
+    except (AttributeError, TypeError):
+        # Module is porbably only partially initialised, so we look at its
+        # spec instead
+        try:
+            orig = abspath(module.__spec__.origin)  # type: ignore
+        except (AttributeError, ValueError, TypeError):
+            orig = None
+
+    if orig is not None and isfile(orig):
+        if orig.endswith(".pyc"):
+            orig = orig[:-1]
+        return orig
+
+    return "<unknown origin>"
+
+
+def _resolve(path):
+    # type: (str) -> Optional[str]
+    """Resolve a (relative) path with respect to sys.path."""
+    for base in sys.path:
+        if isdir(base):
+            resolved_path = abspath(join(base, expanduser(path)))
+            if isfile(resolved_path):
+                return resolved_path
+    return None
+
+
+# Borrowed from the wrapt module
+# https://github.com/GrahamDumpleton/wrapt/blob/0baff1b6ac8d64ffb08ea2d1610c7ed90115b9d5/src/wrapt/importer.py
+class _ImportHookChainedLoader:
+    def __init__(self, loader, callback):
+        # type: (Loader, Callable[[ModuleType], None]) -> None
+        self.loader = loader
+        self.callback = callback
+
+    def load_module(self, fullname):
+        # type: (str) -> ModuleType
+        module = self.loader.load_module(fullname)
+        self.callback(module)
+
+        return module
+
+    def get_code(self, mod_name):
+        return self.loader.get_code(mod_name)
+
+
+class ModuleWatchdog(dict):
+    """Module watchdog.
+
+    Replace the standard ``sys.modules`` dictionary to detect when modules are
+    loaded/unloaded. This is also responsible for triggering any registered
+    import hooks.
+    """
+
+    _run_code = None
+
+    def __init__(self):
+        # type: () -> None
+        self._hook_map = defaultdict(list)  # type: DefaultDict[str, List[Tuple[HookType, Any]]]
+        self._origin_map = {origin(module): module for module in sys.modules.values()}
+        self._modules = sys.modules
+        self._finding = set()  # type: Set[str]
+
+    def __enter__(self):
+        # type: () -> ModuleWatchdog
+        # DEV: We need to patch __import__ because __setitem__ might be getting
+        # called too early and it might be empty or partially initialised.
+        self._patch_import()
+        sys.modules = self
+        log.debug("%s installed", type(self))
+        return self
+
+    def __exit__(self, *exc):
+        # type: (Tuple[Any]) -> None
+        self._unpatch_import()
+        sys.modules = self._modules
+        log.debug("%s uninstalled", type(self))
+
+    def __getitem__(self, item):
+        # type: (str) -> ModuleType
+        return self._modules.__getitem__(item)
+
+    def __setitem__(self, name, module):
+        # type: (str, ModuleType) -> None
+        self._modules.__setitem__(name, module)
+
+    def _patch_import(self):
+        # type: () -> None
+        if sys.meta_path and type(sys.meta_path[0]) is type(self):
```

Comment:
you might have issues with multiple libraries adding import hooks.

you really want to be first, if someone adds a hook in front of you then you might miss some imports.

also, with this check you might add yourself twice.

```
sys.meta_path.insert(0, self)

sys.meta_path.insert(0, some_other_lib)

# since sys.meta_path[0] is not ModuleWatchdog, then you'll insert again
```

Reply:
Ah, good catch! I have fixed this in the `_unpatch_import` because of this very issue, but not here! As for ensuring we stay at the front of `meta_path`, this would be hard to achieve. Normally we would install this as part of the bootstrap process, so anything could happen after that. A potential solution is to wrap around `sys.meta_path`.